core/src/main/java/io/questdb/cutlass/line/http/LineHttpSender.java

```diff
@@ -0,0 +1,637 @@
+                long nowNanos = System.nanoTime();
+                retryingDeadlineNanos = (retryingDeadlineNanos == Long.MIN_VALUE && !closing) ? nowNanos + maxRetriesNanos : retryingDeadlineNanos;
+                if (nowNanos >= retryingDeadlineNanos) {
+                    // we did our best, give up
+                    pendingRows = 0;
+                    request = newRequest();
+                    throw new LineSenderException("Could not flush buffer: ").put(url).put(" Connection Failed").put(": ").put(e.getMessage());
+                }
+                retryBackoff = backoff(retryBackoff);
+            }
+        }
+        pendingRows = 0;
+        request = newRequest();
+    }
+
+    private void handleHttpErrorResponse(DirectUtf8Sequence statusCode, HttpClient.ResponseHeaders response) {
+        // be ready for next request
+        pendingRows = 0;
+        request = newRequest();
+
+        CharSequence statusAscii = statusCode.asAsciiCharSequence();
+        if (Chars.equals("404", statusAscii)) {
+            consumeChunkedResponse(response);
+            client.disconnect();
+            throw new LineSenderException("Could not flush buffer: HTTP endpoint does not support ILP. [http-status=404]");
+        }
+        if (Chars.equals("401", statusAscii) || Chars.equals("403", statusAscii)) {
+            StringSink sink = Misc.getThreadLocalSink();
+            chunkedResponseToSink(response, sink);
+            LineSenderException ex = new LineSenderException("Could not flush buffer: HTTP endpoint authentication error");
+            if (sink.length() > 0) {
+                ex = ex.put(": ").put(sink);
+            }
+            ex.put(" [http-status=").put(statusAscii).put(']');
+            client.disconnect();
+            throw ex;
+        }
+        DirectUtf8Sequence contentType = response.getContentType();
+        if (contentType != null && Chars.equals("application/json", contentType.asAsciiCharSequence())) {
+            if (jsonErrorParser == null) {
+                jsonErrorParser = new JsonErrorParser();
+            }
+            jsonErrorParser.reset();
+            LineSenderException ex = jsonErrorParser.toException(response.getChunkedResponse(), statusCode);
+            client.disconnect();
+            throw ex;
+        }
+        // ok, no JSON, let's do something more generic
+        StringSink sink = Misc.getThreadLocalSink();
+        sink.put("Could not flush buffer: ");
+        chunkedResponseToSink(response, sink);
+        sink.put(" [http-status=").put(statusCode).put(']');
+        client.disconnect();
+        throw new LineSenderException(sink);
+    }
+
+    private boolean isRetryableHttpStatus(DirectUtf8Sequence statusCode) {
+        if (statusCode == null || statusCode.size() != 3 || statusCode.byteAt(0) != '5') {
+            return false;
+        }
+        // we know the status bytes are 5xx, but is the HTTP status retryable?
```

Comment:
Technically speaking, if your request is supposed to be idempotent, any HTTP status is retryable. I'd rather remove this comment as it's confusing and update the status list below to make it more readable.

Reply:
@puzpuzpuz wdyt now?

Reply:
Looks good. Thanks!